src/HealthChecks/HealthChecks/test/HealthCheckPublisherHostedServiceTest.cs

```diff
@@ -239,6 +240,106 @@ public async Task RunAsync_WaitsForCompletion_Single()
+
+        var publishers = new TestPublisher[]
+        {
+                new TestPublisher() { Wait = unblock.Task, },
+        };
+
+        var service = CreateService(publishers, configureBuilder: b =>
+        {
+            b.Add(
+                    new HealthCheckRegistration(
+                        name: "CheckDefault",
+                        instance: new DelegateHealthCheck(_ => Task.FromResult(HealthCheckResult.Healthy(HealthyMessage))),
+                        failureStatus: null,
+                        tags: null));
+
+            b.Add(
+                new HealthCheckRegistration(
+                    name: "CheckDelay2Period18",
+                    instance: new DelegateHealthCheck(_ => Task.FromResult(HealthCheckResult.Healthy(HealthyMessage))),
+                    failureStatus: null,
+                    tags: null,
+                    timeout: default)
+                {
+                    Delay = TimeSpan.FromSeconds(2),
+                    Period = TimeSpan.FromSeconds(18)
+                });
+
+        b.Add(
+                new HealthCheckRegistration(
+                    name: "CheckDelay7Period11",
+                    instance: new DelegateHealthCheck(_ => Task.FromResult(HealthCheckResult.Healthy(HealthyMessage))),
+                    failureStatus: null,
+                    tags: null,
+                    timeout: default)
+                {
+                    Delay = TimeSpan.FromSeconds(7),
+                    Period = TimeSpan.FromSeconds(11)
+                });
+
+            b.Add(
+               new HealthCheckRegistration(
+                   name: "CheckDelay9Period5",
+                   instance: new DelegateHealthCheck(_ =>
+                   {
+                       unblockDelayedCheck.TrySetResult(null); // Unblock last delayed check
+                       return Task.FromResult(HealthCheckResult.Healthy(HealthyMessage));
+                   }),
+                    failureStatus: null,
+                   tags: null,
+                   timeout: default)
+               {
+                   Delay = TimeSpan.FromSeconds(9),
+                   Period = TimeSpan.FromSeconds(5)
+               });
+        });
+
+        try
+        {
+            await service.StartAsync();
+
+            // Act
+            var running = RunServiceAsync(service);
+
+            await publishers[0].Started.TimeoutAfter(TimeSpan.FromSeconds(10));
+
+            unblock.SetResult(null);
+
+            await running.TimeoutAfter(TimeSpan.FromSeconds(20));
+
+            await Task.WhenAll(unblock.Task, unblockDelayedCheck.Task);
+
+            // Assert
+            Assert.True(service.IsTimerRunning);
+            Assert.False(service.IsStopping);
+
+            for (var i = 0; i < publishers.Length; i++)
+            {
+                Assert.True(publishers[i].Entries.Count == 4);
```

Comment:
There's only ever one publisher in this test, right? It's probably simpler to store it in a `publisher` variable and then call `CreateService(new[] { publisher }, ...` I don't think all of ublocking is necessary for this test either. Can it be simplified to the following?

```csharp
            b.Add(
                new HealthCheckRegistration(
                    name: "CheckDelay2Period18v2",
                    instance: new DelegateHealthCheck(_ => Task.FromResult(HealthCheckResult.Healthy(HealthyMessage))),
                    failureStatus: null,
                    tags: null,
                    timeout: default)
                {
                    Delay = TimeSpan.FromSeconds(.2),
                    Period = TimeSpan.FromSeconds(18)
                });
        });

        try
        {
            var running = RunServiceAsync(service);

            await publisher.Started.TimeoutAfter(TimeSpan.FromSeconds(10));

            await Task.Yield();
            Assert.False(running.IsCompleted);

            unblock.SetResult(null);

            await running.TimeoutAfter(TimeSpan.FromSeconds(10));

            // The timer hasn't started yet. Only the default 5 minute registration is run by RunServiceAsync.
            Assert.Equal("CheckDefault", Assert.Single(Assert.Single(publisher.Entries).report.Entries.Keys));

            await service.StartAsync();
            await unblockDelayedCheck.Task.TimeoutAfter(TimeSpan.FromSeconds(10));

            Assert.True(service.IsTimerRunning);
            Assert.False(service.IsStopping);
        }
        finally
        {
            await service.StopAsync();
            Assert.False(service.IsTimerRunning);
            Assert.True(service.IsStopping);
        }

        var entries = publisher.Entries.SelectMany(e => e.report.Entries.Select(e2 => e2.Key)).OrderBy(k => k).ToArray();
        Assert.Equal(
            new[] { "CheckDefault", "CheckDelay2Period18", "CheckDelay2Period18v2", "CheckDelay7Period11", "CheckDelay9Period5" },
            entries);
```

I think having multiple registrations with the same delay in period is worth checking. It's also a good idea to check the entries after the timers have stopped, especially if we're going to have them firing closer together.